File: packages/capacitor-actito/android/src/main/java/com/actito/capacitor/ActitoPlugin.kt
package com.actito.capacitor

import android.content.Intent
import android.net.Uri
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.ktx.device
import com.actito.ktx.events
import com.actito.models.ActitoApplication
import com.actito.models.ActitoDoNotDisturb
import com.actito.models.ActitoDynamicLink
import com.actito.models.ActitoEvent
import com.actito.models.ActitoEventData
import com.actito.models.ActitoNotification
import com.actito.models.ActitoUserData
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject

@CapacitorPlugin(name = "ActitoPlugin")
public class ActitoPlugin : Plugin() {
    override fun load() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)
        Actito.intentReceiver = ActitoPluginIntentReceiver::class.java

        val intent = activity?.intent
        if (intent != null) handleOnNewIntent(intent)
    }

    override fun handleOnNewIntent(intent: Intent) {
        // Try handling the test device intent.
        if (Actito.handleTestDeviceIntent(intent)) return

        // Try handling the dynamic link intent.
        val activity = activity
        if (activity != null && Actito.handleDynamicLinkIntent(activity, intent)) return

        val url = intent.data?.toString()
        if (url != null) {
            EventBroker.dispatchEvent("url_opened", JSObject().apply {
                put("url", url)
            })
        }
    }

    // region Actito

    @PluginMethod
    public fun isConfigured(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.isConfigured)
            }
        )
    }

    @PluginMethod
    public fun isReady(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.isReady)
            }
        )
    }

    @PluginMethod
    public fun launch(call: PluginCall) {
        Actito.launch(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun unlaunch(call: PluginCall) {
        Actito.unlaunch(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun getApplication(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.application?.toJson())
            }
        )
    }

    @PluginMethod
    public fun fetchApplication(call: PluginCall) {
        Actito.fetchApplication(object : ActitoCallback<ActitoApplication> {
            override fun onSuccess(result: ActitoApplication) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchNotification(call: PluginCall) {
        val id = call.getString("id") ?: run {
            call.reject("Missing 'id' parameter.")
            return
        }

        Actito.fetchNotification(id, object : ActitoCallback<ActitoNotification> {
            override fun onSuccess(result: ActitoNotification) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchDynamicLink(call: PluginCall) {
        val url = call.getString("url") ?: run {
            call.reject("Missing 'url' parameter.")
            return
        }

        val uri = Uri.parse(url)

        Actito.fetchDynamicLink(uri, object : ActitoCallback<ActitoDynamicLink> {
            override fun onSuccess(result: ActitoDynamicLink) {
                call.resolve(
                    JSObject().apply {
                        put("result", result.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun canEvaluateDeferredLink(call: PluginCall) {
        Actito.canEvaluateDeferredLink(object : ActitoCallback<Boolean> {
            override fun onSuccess(result: Boolean) {
                call.resolve(
                    JSObject().apply {
                        put("result", result)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun evaluateDeferredLink(call: PluginCall) {
        Actito.evaluateDeferredLink(object : ActitoCallback<Boolean> {
            override fun onSuccess(result: Boolean) {
                call.resolve(
                    JSObject().apply {
                        put("result", result)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    // region Actito device module

    @PluginMethod
    public fun getCurrentDevice(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.device().currentDevice?.toJson())
            }
        )
    }

    @PluginMethod
    public fun getPreferredLanguage(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Actito.device().preferredLanguage)
            }
        )
    }

    @PluginMethod
    public fun updatePreferredLanguage(call: PluginCall) {
        val language = call.getString("language")

        Actito.device().updatePreferredLanguage(language, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun register(call: PluginCall) {
        val userId = call.getString("userId")
        val userName = call.getString("userName")

        Actito.device().register(userId, userName, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun updateUser(call: PluginCall) {
        val userId = call.getString("userId")
        val userName = call.getString("userName")

        Actito.device().updateUser(userId, userName, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchTags(call: PluginCall) {
        Actito.device().fetchTags(object : ActitoCallback<List<String>> {
            override fun onSuccess(result: List<String>) {
                call.resolve(
                    JSObject().apply {
                        put("result", JSArray(result))
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun addTag(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Actito.device().addTag(tag, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun addTags(call: PluginCall) {
        val tags = call.getArray("tags", null)?.toList<String>() ?: run {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Actito.device().addTags(tags, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun removeTag(call: PluginCall) {
        val tag = call.getString("tag") ?: run {
            call.reject("Missing 'tag' parameter.")
            return
        }

        Actito.device().removeTag(tag, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun removeTags(call: PluginCall) {
        val tags = call.getArray("tags", null)?.toList<String>() ?: run {
            call.reject("Missing 'tags' parameter.")
            return
        }

        Actito.device().removeTags(tags, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clearTags(call: PluginCall) {
        Actito.device().clearTags(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchDoNotDisturb(call: PluginCall) {
        Actito.device().fetchDoNotDisturb(object : ActitoCallback<ActitoDoNotDisturb?> {
            override fun onSuccess(result: ActitoDoNotDisturb?) {
                call.resolve(
                    JSObject().apply {
                        put("result", result?.toJson())
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun updateDoNotDisturb(call: PluginCall) {
        val dnd = call.getObject("dnd", null)?.let { ActitoDoNotDisturb.fromJson(it) } ?: run {
            call.reject("Missing 'dnd' parameter.")
            return
        }

        Actito.device().updateDoNotDisturb(dnd, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun clearDoNotDisturb(call: PluginCall) {
        Actito.device().clearDoNotDisturb(object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchUserData(call: PluginCall) {
        Actito.device().fetchUserData(object : ActitoCallback<ActitoUserData> {
            override fun onSuccess(result: ActitoUserData) {
                val userData = JSONObject().apply {
                    result.forEach {
                        put(it.key, it.value)
                    }
                }

                call.resolve(
                    JSObject().apply {
                        put("result", userData)
                    }
                )
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun updateUserData(call: PluginCall) {
        val json = call.getObject("userData", null) ?: run {
            call.reject("Missing 'userData' parameter.")
            return
        }

        val userData = mutableMapOf<String, String?>().apply {
            val iterator = json.keys()
            while (iterator.hasNext()) {
                val key = iterator.next()
                val value = json.getString(key)

                put(key, value)
            }
        }

        Actito.device().updateUserData(userData, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion

    // region Actito events module

    @PluginMethod
    public fun logCustom(call: PluginCall) {
        val event = call.getString("event") ?: run {
            call.reject("Missing 'event' parameter.")
            return
        }

        val data: ActitoEventData?

        try {
            data = call.getObject("data", null)?.let { ActitoEvent.createData(it) }
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.events().logCustom(event, data, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    // endregion
}

File: packages/capacitor-actito/ios/Sources/ActitoPlugin/ActitoPlugin.swift
import ActitoKit
import Capacitor
import Foundation

@objc(ActitoPlugin)
public class ActitoPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ActitoPlugin"
    public let jsName = "ActitoPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isConfigured", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isReady", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "launch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unlaunch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getApplication", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchApplication", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchNotification", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchDynamicLink", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "canEvaluateDeferredLink", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "evaluateDeferredLink", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentDevice", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPreferredLanguage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updatePreferredLanguage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "register", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addTag", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeTag", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchDoNotDisturb", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateDoNotDisturb", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearDoNotDisturb", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fetchUserData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateUserData", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logCustom", returnType: CAPPluginReturnPromise),
    ]

    @MainActor
    public override func load() {
        addApplicationLaunchListener()

        EventBroker.instance.setup { self.notifyListeners($0, data: $1, retainUntilConsumed: $2) }
        Actito.shared.delegate = self

        NotificationCenter.default.addObserver(self, selector: #selector(self.handleUrlOpened(notification:)), name: .capacitorOpenURL, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.handleUniversalLink(notification:)), name: .capacitorOpenUniversalLink, object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self, name: .capacitorOpenURL, object: nil)
        NotificationCenter.default.removeObserver(self, name: .capacitorOpenUniversalLink, object: nil)
    }

    @objc func handleUrlOpened(notification: NSNotification) {
        guard let object = notification.object as? [String: Any?], let url = object["url"] as? URL else {
            logger.warning("Unprocessable url_opened event.")
            return
        }

        DispatchQueue.main.async {
            if Actito.shared.handleTestDeviceUrl(url) {
                return
            }

            if Actito.shared.handleDynamicLinkUrl(url) {
                return
            }

            EventBroker.instance.dispatchEvent("url_opened", data: ["url": url.absoluteString], retainUntilConsumed: true)
        }
    }

    @objc func handleUniversalLink(notification: NSNotification) {
        guard let object = notification.object as? [String: Any?], let url = object["url"] as? URL else {
            logger.warning("Unprocessable universal_link event.")
            return
        }

        DispatchQueue.main.async {
            if Actito.shared.handleTestDeviceUrl(url) {
                return
            }

            _ = Actito.shared.handleDynamicLinkUrl(url)
        }
    }

    // MARK: - Actito

    @objc func isConfigured(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "result": Actito.shared.isConfigured,
            ])
        }
    }

    @objc func isReady(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "result": Actito.shared.isReady,
            ])
        }
    }

    @objc func launch(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.launch { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func unlaunch(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.unlaunch { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func getApplication(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                var response: PluginCallResultData = [:]
                if let application = Actito.shared.application {
                    response["result"] = try application.toJson()
                }

                call.resolve(response)
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchApplication(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.fetchApplication { result in
                switch result {
                case let .success(application):
                    do {
                        call.resolve([
                            "result": try application.toJson()
                        ])
                    } catch {
                        call.reject(error.localizedDescription)
                    }
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchNotification(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("Missing 'id' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.fetchNotification(id) { result in
                switch result {
                case let .success(notification):
                    do {
                        call.resolve([
                            "result": try notification.toJson()
                        ])
                    } catch {
                        call.reject(error.localizedDescription)
                    }
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchDynamicLink(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("Missing 'url' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.fetchDynamicLink(url) { result in
                switch result {
                case let .success(dynamicLink):
                    do {
                        call.resolve([
                            "result": try dynamicLink.toJson()
                        ])
                    } catch {
                        call.reject(error.localizedDescription)
                    }
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func canEvaluateDeferredLink(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "result": Actito.shared.canEvaluateDeferredLink
            ])
        }
    }

    @objc func evaluateDeferredLink(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.evaluateDeferredLink { result in
                switch result {
                case let .success(evaluated):
                    call.resolve([
                        "result": evaluated
                    ])
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    // MARK: - Actito device module

    @objc func getCurrentDevice(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                var response: PluginCallResultData = [:]
                if let device = Actito.shared.device().currentDevice {
                    response["result"] = try device.toJson()
                }

                call.resolve(response)
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func getPreferredLanguage(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            var response: PluginCallResultData = [:]
            if let language = Actito.shared.device().preferredLanguage {
                response["result"] = language
            }

            call.resolve(response)
        }
    }

    @objc func updatePreferredLanguage(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let language = call.getString("language")

            Actito.shared.device().updatePreferredLanguage(language) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func register(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let userName = call.getString("userName")

        DispatchQueue.main.async {
            Actito.shared.device().register(userId: userId, userName: userName) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func updateUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let userName = call.getString("userName")

        DispatchQueue.main.async {
            Actito.shared.device().updateUser(userId: userId, userName: userName) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchTags(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.device().fetchTags { result in
                switch result {
                case let .success(tags):
                    call.resolve([
                        "result": tags
                    ])
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func addTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.device().addTag(tag) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func addTags(_ call: CAPPluginCall) {
        guard let tags = call.getArray("tags", String.self) else {
            call.reject("Missing 'tags' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.device().addTags(tags) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func removeTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("Missing 'tag' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.device().removeTag(tag) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func removeTags(_ call: CAPPluginCall) {
        guard let tags = call.getArray("tags", String.self) else {
            call.reject("Missing 'tags' parameter.")
            return
        }

        DispatchQueue.main.async {
            Actito.shared.device().removeTags(tags) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func clearTags(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.device().clearTags { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchDoNotDisturb(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.device().fetchDoNotDisturb { result in
                switch result {
                case let .success(dnd):
                    do {
                        var response: PluginCallResultData = [:]
                        if let dnd = dnd {
                            response["result"] = try dnd.toJson()
                        }

                        call.resolve(response)
                    } catch {
                        call.reject(error.localizedDescription)
                    }
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func updateDoNotDisturb(_ call: CAPPluginCall) {
        guard let json = call.getObject("dnd") else {
            call.reject("Missing 'dnd' parameter.")
            return
        }

        let dnd: ActitoDoNotDisturb
        do {
            dnd = try ActitoDoNotDisturb.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        DispatchQueue.main.async {
            Actito.shared.device().updateDoNotDisturb(dnd) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func clearDoNotDisturb(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.device().clearDoNotDisturb { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func fetchUserData(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Actito.shared.device().fetchUserData { result in
                switch result {
                case let .success(userData):
                    call.resolve([
                        "result": userData
                    ])
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    @objc func updateUserData(_ call: CAPPluginCall) {
        guard let json = call.getObject("userData") else {
            call.reject("Missing 'userData' parameter.")
            return
        }

        let userData = json.mapValues { $0 as? String }

        DispatchQueue.main.async {
            Actito.shared.device().updateUserData(userData) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }

    // MARK: - Actito events module

    @objc func logCustom(_ call: CAPPluginCall) {
        guard let event = call.getString("event") else {
            call.reject("Missing 'event' parameter.")
            return
        }

        let data = call.getObject("data")

        DispatchQueue.main.async {
            Actito.shared.events().logCustom(event, data: data) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
        }
    }
}

extension ActitoPlugin: ActitoDelegate {
    public func actito(_ actito: Actito, onReady application: ActitoApplication) {
        do {
            EventBroker.instance.dispatchEvent("ready", data: try application.toJson())
        } catch {
            logger.error("Failed to emit the ready event.", error: error)
        }
    }

    public func actitoDidUnlaunch(_ actito: Actito) {
        EventBroker.instance.dispatchEvent("unlaunched", data: nil)
    }

    public func actito(_ actito: Actito, didRegisterDevice device: ActitoDevice) {
        do {
            EventBroker.instance.dispatchEvent("device_registered", data: try device.toJson())
        } catch {
            logger.error("Failed to emit the device_registered event.", error: error)
        }
    }
}

extension ActitoPlugin {
    private func addApplicationLaunchListener() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(didFinishLaunching),
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    private func removeApplicationLaunchListener() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didFinishLaunchingNotification,
            object: nil
        )
    }

    @MainActor
    @objc private func didFinishLaunching() {
        removeApplicationLaunchListener()

        logger.hasDebugLoggingEnabled = Actito.shared.options?.debugLoggingEnabled ?? false
    }
}
